### PR TITLE
52892: Privacy Export Personal Data: JSON encoding failure generates invalid JSON in export file

### PR DESCRIPTION
**Pull request by hellofromtonya**
Trac ticket: https://core.trac.wordpress.org/ticket/52892

Function: `wp_privacy_generate_personal_data_export_file()`

PHPStan flagged a type mismatch if `false` is passed into `fwrite` for groups JSON:
```php
fwrite( $file, $groups_json );
```

The groups JSON is assigned from `wp_json_encode`:
```php
$groups_json = wp_json_encode( $groups );
```

The [return signatures of `wp_json_encode`](https://developer.wordpress.org/reference/functions/wp_json_encode/#return) are `string` or `false`:
>_(string|false)_ The JSON encoded string, or false if it cannot be encoded.

### Is Passing `false` a problem?
See it in action here https://3v4l.org/bpmqG

- No errors, warnings, or notices happen in PHP 5.6 to 8.0.3
- Generates invalid JSON as `false` writes nothing into the file:
```json
"{Personal Data Export for tester@test.com:}"
```

This is the current behavior 👆, where generating invalid JSON is a bug.

### What can cause JSON encoding to fail?
`json_encode` returns `false` when:

- attempting to encode a non-UTF-8 string
- a resource is given

WordPress automatically remedies the encoding with `_wp_json_sanity_check` where it converts string(s) to UTF-8.

A resource could exist an element of the groups _when_ the post meta data is filtered. Using this [test script](https://gist.github.com/hellofromtonya/5f8fc3cfb8c51e448d01d8c72d0d291c) reproduces the failed JSON encoding and invalid JSON report.

## Bug fix solution

Update: After pair programming session, @jrfnl and I agreed this is a bug. Fixing the JSON solves the invalid JSON problem; however, the export is empty and not usable. Instead, generating an error alerts that there 's a problem, which is a preferred experience.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php [hunk 1]
```diff
@@ -94,6 +94,7 @@ public function setUp() {
 	 * @since 5.2.0
 	 */
 	public function tearDown() {
+		remove_filter( 'get_post_metadata', array( $this, 'filter_export_data_grouped_metadata' ) );
```
Comment (hellofromtonya) [inline tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php:97]:
```suggestion
		// Unhook for test_should_generate_valid_json_when_json_encoding_fails.
		remove_filter( 'get_post_metadata', array( $this, 'filter_export_data_grouped_metadata' ) );
```

File: tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php [hunk 2]
```diff
@@ -94,6 +94,9 @@ public function setUp() {
 	 * @since 5.2.0
 	 */
 	public function tearDown() {
+		// Reset for test_should_generate_valid_json_when_json_encoding_fails.
```
Comment (hellofromtonya) [inline tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php:97]:
```suggestion
		// Reset for test_should_generate_valid_json_when_json_encoding_fails.
```

File: src/wp-admin/includes/privacy-tools.php [hunk 3]
```diff
@@ -400,6 +400,17 @@ function wp_privacy_generate_personal_data_export_file( $request_id ) {
 
 	// Convert the groups to JSON format.
 	$groups_json = wp_json_encode( $groups );
+	if ( false === $groups_json ) {
+		$groups_json = '"false"';
```
Comment (hellofromtonya) [inline src/wp-admin/includes/privacy-tools.php:404]: Change to `'"null"'` for consistency as a value here adds no value nor is descriptive to indicate the cause that generated it.

Comment (hellofromtonya) [reply src/wp-admin/includes/privacy-tools.php]: Resolved in commit https://github.com/WordPress/wordpress-develop/pull/1115/commits/8c6f37f1c2f845e485befa07668bd95ccc4d10c1

File: tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php [hunk 4]
```diff
@@ -654,4 +656,65 @@ public function test_multiple_group_export_single_items_no_group_count_in_toc()
 		$this->assertContains( $request->email, $report_contents );
 
 	}
+
+	/**
+	 * Test should generate a valid JSON report when/if JSON encoding fails.
+	 *
+	 * @ticket 51423
+	 */
+	public function test_should_generate_valid_json_when_json_encoding_fails() {
```
Comment (hellofromtonya) [inline tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php:665]: Update tests for the latest code change to catch the JSON error and check the message. No longer to open the file and check the JSON.

Comment (hellofromtonya) [reply tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php]: Resolved in commit https://github.com/WordPress/wordpress-develop/pull/1115/commits/368792134756fb448864290a5985643874bf2894

**Review by jrfnl — APPROVED** [review]

**Comment by hellofromtonya** [comment]
Committed with changeset https://core.trac.wordpress.org/changeset/50713

**Pull request closed without merging**